### PR TITLE
254 notification comments

### LINKED ISSUE #254: As a user, I can comment on a booking
### User story
As a user, I can comment on a booking. 

### Narrative
Comment system for booking requests. Users should be notified when there are comments on their bookings, and approvers should be notified of new comments on bookings they have previously commented on. 

Comments should use a Rich-Text editor

The scope includes all necessary changes (frontend, backend, tests)

### Out of scope
Scope is limited to basic comment system (no replies, mentions etc)

### Acceptance criteria

| # | Given | When | Then |
|---|-------|------|------|

### PR DESCRIPTION
**Pull request by danseremet**
##### **_IMPORTANT: Always create an issue before a PR and provide enough information so that others can review_**

# Notify reviewers when a user makes a comment.

image

image

##### _List all issues this PR resolves_
- closes #254, #290

##### _Complete these items before requesting reviews_
- [x] Static Analysis Rules Validated
- [x] All commits follow naming conventions

### Sub-tasks Checklist
- [x] UI Implementation
- [x] UI Test
- [x] Backend Implementation
- [x] Backend Tests
- [x] Integration Tests
- [x] System Tests
- [x] Documentation

### CODE REVIEW

File: app/Notifications/CommentNotification.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+     * Get the mail representation of the notification.
+     *
+     * @param mixed $notifiable
+     * @return \Illuminate\Notifications\Messages\MailMessage
+     */
+    public function toMail($notifiable)
+    {
+        return (new MailMessage)
+            ->line('There are new comments on a booking.')
+            ->action('View Review', url('/bookings/' . $this->comment->booking_id . '/review'))
```
Comment (a-drew) [inline app/Notifications/CommentNotification.php:47]:
```suggestion
            ->action('View Review', route('bookings.review' , $this->comment->booking_id))
```

File: app/Notifications/CommentNotification.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+     * @param mixed $notifiable
+     * @return array
+     */
+    public function toArray($notifiable)
+    {
+        return [
+            //
+        ];
+    }
+}
```
Comment (jonathanp16) [inline app/Notifications/CommentNotification.php:63]: Do we need this or is it just boilerplate code

**Review by jonathanp16 — DISMISSED** [review]

File: tests/Browser/BookingReviewTest.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                ->type('.ProseMirror', 'This is a test comment.')
+                ->pressAndWaitFor('@saveText')
+                ->releaseMouse()
+                ->scrollTo('.ProseMirror')
+                ->assertSee("This is a test comment.");
+        });
+
+        $this->assertDatabaseHas('email_log', [
+            'subject' => __('Comment Notification')
+        ]);
```
Comment (a-drew) [inline tests/Browser/BookingReviewTest.php:73]: No need to import a package. you can mock mail and notifcations : https://laravel.com/docs/8.x/mocking#notification-fake

```
class ExampleTest extends TestCase
{
    public function test_comment_send_notification()
    {
        Notification::fake();

        // Assert that no notifications were sent...
        Notification::assertNothingSent();
        
        // Perform dusk browsers stuff

        // Assert a notification was sent to the given users...
        Notification::assertSentTo(
            [$user], CommentNotification::class
        );

        // Assert a notification was not sent...
        Notification::assertNotSentTo(
            [$user], CommentNotification::class
        );
    }
}
```

File: app/Notifications/CommentNotification.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+     *
+     * @param mixed $notifiable
+     * @return array
+     */
+    public function toArray($notifiable)
+    {
+        return [
+            //
+        ];
+    }
```
Comment (alexstojda) [inline app/Notifications/CommentNotification.php:62]: This should contain the "object" form of the notification. 
```suggestion
    public function toArray($notifiable)
    {
        return [
            'message' => "" //Add whatever message 
            'link' => url('/bookings/' . $this->comment->booking_id . '/review')
            // Add more context info here as requested in my other review 
        ];
    }
```

File: app/Notifications/CommentNotification.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * Get the mail representation of the notification.
+     *
+     * @param mixed $notifiable
+     * @return \Illuminate\Notifications\Messages\MailMessage
+     */
+    public function toMail($notifiable)
+    {
+        return (new MailMessage)
+            ->line('There are new comments on a booking.')
```
Comment (alexstojda) [inline app/Notifications/CommentNotification.php:46]: Add more context to the notification. What booking is this for? Why are they receiving this notification (Are they a reviewer? the submitter?) ? 

If a user has many bookings, this email notification is not useful or informative.

File: app/Observers/CommentObserver.php [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+class CommentObserver
+{
+    public function saved(Comment $comment)
+    {
+        if(Auth::check())
+        {
+            $booking = $comment->booking()->first();
+            $reviewers = $booking->reviewers()->get();
+            foreach ($reviewers as $reviewer) {
+                $reviewer->notify(new CommentNotification($comment));
```
Comment (alexstojda) [inline app/Observers/CommentObserver.php:18]: probably a good idea to pass the booking to the comment for the context info mentioned above

**Review by alexstojda — CHANGES_REQUESTED** [review]
A few things to address

File: tests/Browser/BookingReviewTest.php (hunk 3 above)
Comment (alexstojda) [reply tests/Browser/BookingReviewTest.php:73]: @a-drew that only works for feature & unit tests unfortunately.

The dusk test runner is a separate process from the app, so it has no access to intercept & mock calls

File: tests/Browser/BookingReviewTest.php [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            ->count(1)
+            ->hasReservations(random_int(1, 3))
+            ->create(["status" => BookingRequest::REVIEW])->first();
+
+        $this->browse(function (Browser $browser) use ($booking)
+        {
+            $browser->loginAs(User::first());
+            $browser->visit('/bookings/review')
+                ->clickLink('Open Details')
+                ->pause(5000)
```
Comment (alexstojda) [inline tests/Browser/BookingReviewTest.php:41]:
```suggestion
                ->pause(250)
```

As mentioned in other PRs, avoid long pauses unless absolutely necessary.

File: tests/Browser/BookingReviewTest.php [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
```
Comment (alexstojda) [inline tests/Browser/BookingReviewTest.php:59]:
```suggestion
                ->pause(250)
```

**Review by alexstojda — APPROVED** [review]

**Review by alexstojda — CHANGES_REQUESTED** [review]
Oops, wrong  button before lol

File: tests/Browser/BookingReviewTest.php (hunk 7 above)
Comment (a-drew) [reply tests/Browser/BookingReviewTest.php]: I'd actually say use `$browser->waitFor...()` style helpers instead. less total wait time and still solid on ci.

File: tests/Browser/BookingReviewTest.php (hunk 3 above)
Comment (a-drew) [reply tests/Browser/BookingReviewTest.php:73]: ahhh I gotcha. ngl testing for notifications should've been a backend test, much faster. 
But with dusk you get a 2 for one kind of deal so 🤷

for any future facade related needs: https://github.com/NoelDeMartin/laravel-dusk-mocking

File: tests/Browser/BookingReviewTest.php (hunk 7 above)
Comment (alexstojda) [reply tests/Browser/BookingReviewTest.php]: Good call, @danseremet if you can use a waitFor instead that would be best

File: app/Notifications/CommentNotification.php (hunk 2 above)
Comment (danseremet) [reply app/Notifications/CommentNotification.php]: boilerplate

File: app/Notifications/CommentNotification.php (hunk 1 above)
Comment (danseremet) [reply app/Notifications/CommentNotification.php]: done

File: app/Notifications/CommentNotification.php (hunk 4 above)
Comment (danseremet) [reply app/Notifications/CommentNotification.php]: as we discussed, not needed anymore

File: app/Observers/CommentObserver.php (hunk 6 above)
Comment (danseremet) [reply app/Observers/CommentObserver.php]: done

File: tests/Browser/BookingReviewTest.php (hunk 7 above)
Comment (danseremet) [reply tests/Browser/BookingReviewTest.php]: done

File: app/Notifications/CommentNotification.php (hunk 5 above)
Comment (danseremet) [reply app/Notifications/CommentNotification.php]: done

**Review by a-drew — APPROVED** [review]

**Review by a-drew — DISMISSED** [review]
no fail pls

**Review by alexstojda — APPROVED** [review]

**Review by EvanDime — APPROVED** [review]

**Pull request merged**